string[] locations
int32 loop
---
string message
